### ISSUE TITLE
Few Questions about sbert cross encoder & CEBinaryClassificationEvaluator

### ISSUE THREAD

**Issue by duarteharris** [UKPLab/sentence-transformers#2624]
First of all, let me share my appreciation for this library. It's been great to use so far and I really appreciate it.

I've been looking at the [Cross Encoders section](https://www.sbert.net/examples/applications/cross-encoder/README.html) and used the CrossEncoder example for the [Quora Duplicate Questions Detection task](https://github.com/UKPLab/sentence-transformers/blob/master/examples/training/cross-encoder/training_quora_duplicate_questions.py) to build my own model (which shows great eval metrics, thanks!).

A few questions emerged and I haven't been able to find answers for them:

1. During training, using the CEBinaryClassificationEvaluator we see something like:

CEBinaryClassificationEvaluator: Evaluating the model on dev dataset in epoch 1 after 60000 steps:
Accuracy:           97.15	(Threshold: 0.8388)
F1:                 84.52	(Threshold: 0.7532)
Precision:          86.67
Recall:             82.48
Average Precision:  91.55

What are these Thresholds we see for the Accuracy and the F1 Score? What do they mean? In my case, for the F1 Score they seem to go all over the place (in one it can show the 0.7532 you see there and the next show 0.1102 and just keep jumping around). What information should I glean from this?

2. What is the training optimising for? Accuracy? Loss? Something else? Is there a way to choose?

3. Can we choose other metrics, such as Fbeta Score, Matthews Correlation Coefficient  and/or Error Rate and add  them to the eval?

4. On the training of the CrossEncoders it reads: "For binary tasks and tasks with continuous scores (like STS), we set num_labels=1. For classification tasks, we set it to the number of labels we have.". So: if I'm understanding correctly, if I want the model.predict to return not a value between 0 and 1, but rather the integer 0 or 1 (in case of a binary problem) I need to set up num_labels=2, correct? 

Thank you for your attention in advance and thank you once again for the library,

**Comment by ir2718** [comment]
Hi,

I'm going to try to answer your questions as best as I can.

> What are these Thresholds we see for the Accuracy and the F1 Score? What do they mean? In my case, for the F1 Score they seem to go all over the place (in one it can show the 0.7532 you see there and the next show 0.1102 and just keep jumping around). What information should I glean from this?

The QQP dataset is a classification dataset, with labels being either 0 or 1. In case you have two classes the usual approach is setting `num_labels` to 1 (exactly as in the example you linked), meaning `BCEWithLogitsLoss` will be used. Under the hood, the `CrossEncoder` uses a sigmoid activation function on the logits, which outputs probabilities between 0 and 1. In the end, you just want the predictions, so you can further optimize the metric you want by changing the threshold eg. if the output probability is 0.654, while the label is 0, you can set the threshold to 0.7, which will result in correct classification. 

> What is the training optimising for? Accuracy? Loss? Something else? Is there a way to choose?

Since you cannot directly optimize a metric, the `CrossEncoder.fit` method optimizes the loss (in your case `BCEWithLogitsLoss`, while the metric that's indirectly optimized is average precision. Currently, there is no way to choose this, but you can fork the library and just return the metric that you want to optimize for in the `CEBinaryClassificationEvaluator.__call__` function (since the function already calculates F1, accuracy, precision and recall).

> Can we choose other metrics, such as Fbeta Score, Matthews Correlation Coefficient and/or Error Rate and add them to the eval?

You can do that, but you will have to fork the library and add the calculation of these metrics to the `CEBinaryClassificationEvaluator.__call__` function.

> On the training of the CrossEncoders it reads: "For binary tasks and tasks with continuous scores (like STS), we set num_labels=1. For classification tasks, we set it to the number of labels we have.". So: if I'm understanding correctly, if I want the model.predict to return not a value between 0 and 1, but rather the integer 0 or 1 (in case of a binary problem) I need to set up num_labels=2, correct?

Without going into too much theory, in the case of having two labels, you can set `num_labels` to 1, because the sum of probabilities will always be 1.

In the case you do set `num_labels` to 2 the output of the `CrossEncoder.predict` method will be the resulting logits, but you can use the `apply_softmax` parameter to get probabilities. If you just want integer values, you will have to use an argmax operation on the final tensor.

I hope this answers all your questions. If I left something out feel free to point it out.

**Comment by duarteharris** [comment]
Hi @ir2718 . Thank you for the response. Hope you don't mind clarifying some points as they're still not clear to me.

> The QQP dataset is a classification dataset, with labels being either 0 or 1. In case you have two classes the usual approach is setting num_labels to 1 (exactly as in the example you linked), meaning BCEWithLogitsLoss will be used. Under the hood, the CrossEncoder uses a sigmoid activation function on the logits, which outputs probabilities between 0 and 1. In the end, you just want the predictions, so you can further optimize the metric you want by changing the threshold eg. if the output probability is 0.654, while the label is 0, you can set the threshold to 0.7, which will result in correct classification.

Maybe I'm missing something, but this doesn't seem to explain what's the meaning of the Accuracy and F1 Thresholds that are reported when using the CEBinaryClassificationEvaluator. Nor does it explain why it keeps wildly jumping around seemingly without any connection to the Accuracy and F1 scores.

Let's say that on step 5_000 we have:

```
Accuracy: 96.25 (Threshold: 0.3187)
F1: 79.39 (Threshold: 0.3138)
```

then on step 10_000 we have:

```
Accuracy: 96.33 (Threshold: 0.8399)
F1: 79.79 (Threshold: 0.7481)

```
then on step 15_000 we have:

```
Accuracy: 96.38 (Threshold: 0.8025)
F1: 80.45 (Threshold: 0.1212)
```

And we finish with:

```
Accuracy: 98.10 (Threshold: 0.9236)
F1: 80.45 (Threshold: 0.8435)
```

Are you suggesting that Threshold that we should use if we only want the prediction is positive when over 0.8435 (otherwise it's the negative class), but if it had stoped on the previous one it would've been 0.1212..? This doesn't seem right, and I'm using actual numbers.

If not, what is it that you're saying? And how are you choosing and changing the threshold?

> Since you cannot directly optimize a metric, the CrossEncoder.fit method optimizes the loss (in your case BCEWithLogitsLoss), while the metric that's indirectly optimized is average precision. Currently, you can choose the loss function for optimization, but there is no (easy) way to choose the metric that's optimized. If you want to change the metric that's optimized you can fork the library and just return the metric that you want to optimize for in the CEBinaryClassificationEvaluator.__call__ function (since the function already calculates F1, accuracy, precision and recall).

Since it's optimising for loss, is there a way to check what's the loss/ the loss progress as it trains?

> In the case you do set num_labels to 2 the output of the CrossEncoder.predict method will be the resulting logits, but you can use the apply_softmax parameter to get probabilities. If you just want integer values, you will have to use an argmax operation on the final tensor. The result will be the same regardless of setting apply_softmax to True as it preserves ranks.

you mean something like model.predict(..., softmax= True)?


Everything else's clear. And once again, thank you for your answers and patience.

**Comment by ir2718** [comment]
@duarteharris 

> Maybe I'm missing something, but this doesn't seem to explain what's the meaning of the Accuracy and F1 Thresholds that are reported when using the CEBinaryClassificationEvaluator. Nor does it explain why it keeps wildly jumping around seemingly without any connection to the Accuracy and F1 scores.

> And how are you choosing and changing the threshold?

**Short answer**: 
The accuracy and F1 score that's printed in your console when training the model is the maximum value of accuracy/F1. This maximum value is found by sorting the probabilities (and corresponding labels) and then changing the value of the threshold to the mean of two neighbor probability values and evaluating accuracy/F1 score on the validation set. There are no "rules" for how the threshold will behave over the epochs. The only guarantee is that it will always be between 0 and 1.

**Long answer**:
Consider these values as the sorted logits a model predicted and corresponding labels:
```
import numpy as np
labels = np.array([0,   0,   0,   1,   1,   1])
logits = np.array([-1.38629436, -0.84729786,  0.40546511,  0.84729786,  1.38629436, 2.19722458])
```
This means that applying the sigmoid function on the logits will result in probabilities (values defined between 0 and 1).
```
probs = 1 / (1 + np.exp(-logits))
print(probs)                           # array([0.2, 0.3, 0.6, 0.7, 0.8, 0.9])
```
In order to acquire predictions, the usual approach is saying all values higher than 0.5 are 1, all those below 0.5 are 0:
```
preds = (probs > 0.5).astype(int)
print(preds)                           # array([0, 0, 1, 1, 1, 1])
```
Now if you try to calculate the accuracy and F1 score you would get these results:
```
from sklearn.metrics import accuracy_score, f1_score
print(accuracy_score(labels, preds))   # 0.8333333333333334
print(f1_score(labels, preds))         # 0.8571428571428571
```

But, nothing is stopping you from modifying this threshold, as the `CrossEncoder` outputs probabilities in the case of `num_labels=1`. Finding the best threshold would mean trying out these values (means of neighboring sorted probabilities): 0.25, 0.45, 0.65, 0.75, 0.85 

The maximum accuracy and F1 score is achieved by setting the threshold value to 0.65:
 ```
preds = (probs > 0.65).astype(int)
print(preds)                            # array([0, 0, 0, 1, 1, 1])
print(accuracy_score(labels, preds))    # 1.0
print(f1_score(labels, preds))          # 1.0
```

> Are you suggesting that Threshold that we should use if we only want the prediction is positive when over 0.8435 (otherwise it's the negative class), but if it had stoped on the previous one it would've been 0.1212..? This doesn't seem right, and I'm using actual numbers.

If the training goes on for eg. 10 epochs, the final saved model will be the one with the highest average precision (if you don't modify the library). For the sake of example let's say the highest average precision is achieved for epoch 6. Then, you should find the corresponding logged/saved threshold values for epoch 6 and use those if you want to maximize the accuracy/F1 score WHILE having the highest average precision on unseen data. Keep in mind that this might not be the best accuracy/F1 achieved throughout all epochs.

Again, another option is to modify the `CEBinaryClassificationEvaluator.__call__` function to return the accuracy/F1 score. That way the final saved model will be the one with the best accuracy/F1 score. Then you can again find the corresponding logged/saved threshold values and use those for predictions on unseen data.  This approach will guarantee the best accuracy/F1 achieved throughout all epochs.

> Since it's optimising for loss, is there a way to check what's the loss/ the loss progress as it trains?

Achieving the lowest validation loss does not necessarily mean you will achieve the highest metric value. But to answer your question, this currently isn't possible, unless you want to modify it yourself.

> you mean something like model.predict(..., softmax= True)?

I mean `model.predict([["This is some text"]], apply_softmax=True)`. Check the [docs](https://www.sbert.net/docs/package_reference/cross_encoder.html#sentence_transformers.cross_encoder.CrossEncoder.predict).
```
def predict(
        self,
        sentences: List[List[str]],
        batch_size: int = 32,
        show_progress_bar: bool = None,
        num_workers: int = 0,
        activation_fct=None,
        apply_softmax=False,
        convert_to_numpy: bool = True,
        convert_to_tensor: bool = False,
    ):
```

**Comment by duarteharris** [comment]
Thank you once again for your time and patience @ir2718. It's been very helpful. I truly appreciate it.

**Comment by duarteharris** [comment]
Thank you once again for your time and patience @ir2718. It's been very helpful. I truly appreciate it.
